### PR TITLE
net: lwm2m: Way to pass a destination hostname to socket

### PR DESCRIPTION
**Pull request by icatorze**
net: lwm2m: When mbedtls CONFIG_MBEDTLS_SERVER_NAME_INDICATION is
enabled, a destination hostname must be passed to socket to properly
connect do lwm2m server.

Passing lwm2m context to lwm2m_parse_peerinfo

Signed-off-by: Jair Jack <jack@icatorze.com.br>

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 1]
```diff
@@ -144,6 +144,7 @@ const char *lwm2m_engine_get_attr_name(const struct lwm2m_attr *attr);
 int  lwm2m_socket_add(struct lwm2m_ctx *ctx);
 void lwm2m_socket_del(struct lwm2m_ctx *ctx);
 int  lwm2m_socket_start(struct lwm2m_ctx *client_ctx);
-int  lwm2m_parse_peerinfo(char *url, struct sockaddr *addr, bool *use_dtls, bool is_firmware_uri);
+int  lwm2m_parse_peerinfo(char *url, struct lwm2m_ctx *client_ctx,
+							bool is_firmware_uri);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:148]: This should fit into a single line, no?

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 2]
```diff
@@ -4718,7 +4730,8 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 	return lwm2m_socket_add(client_ctx);
 }
 
-int lwm2m_parse_peerinfo(char *url, struct sockaddr *addr, bool *use_dtls, bool is_firmware_uri)
+int lwm2m_parse_peerinfo(char *url, struct lwm2m_ctx *client_ctx,
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4733]: Same here.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 3]
```diff
@@ -4699,6 +4700,17 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
...
 			return -errno;
 		}
+
+		if (client_ctx->desthostname != NULL &&
+				client_ctx->desthostnamelen <= URI_MAX_LEN) {
+			ret = setsockopt(client_ctx->sock_fd, SOL_TLS, TLS_HOSTNAME,
+					client_ctx->desthostname, client_ctx->desthostnamelen);
+			if (ret < 0) {
+				LOG_ERR("Failed to set TLS_HOSTNAME option: %d",
+								errno);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4710]: And here.

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 4]
```diff
@@ -4699,6 +4700,17 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 			lwm2m_engine_context_close(client_ctx);
 			return -errno;
 		}
+
+		if (client_ctx->desthostname != NULL &&
+				client_ctx->desthostnamelen <= URI_MAX_LEN) {
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:4705]: I don't think it's needed to verify the length, socket implementation does not put a limit on that (as long as there's enough mem on heap, but that we are not able to verify here).

File: include/net/lwm2m.h [hunk 5]
```diff
@@ -94,10 +94,17 @@ struct lwm2m_ctx {
...
+	 *  hostname server.
+	 */
+	char *desthostname;
+	uint16_t desthostnamelen;
+
 	/** Client can set load_credentials function as a way of overriding
 	 *  the default behavior of load_tls_credential() in lwm2m_engine.c
 	 */
 	int (*load_credentials)(struct lwm2m_ctx *client_ctx);
+
```
Comment (rlubos) [inline include/net/lwm2m.h:107]: This empty line is not needed.

File: include/net/lwm2m.h [hunk 6]
```diff
@@ -94,10 +94,17 @@ struct lwm2m_ctx {
 	 */
 	int tls_tag;
 
+	/** When MBEDTLS SNI is enabled socket must be set with destination
+	 *  hostname server.
+	 */
+	char *desthostname;
+	uint16_t desthostnamelen;
```
Comment (rlubos) [inline include/net/lwm2m.h:101]: Do I miss something or are `desthostname` and `desthostnamelen` not set in the parser funcition?

Comment (icatorze) [reply include/net/lwm2m.h:101]: now we need set it at application code (like lwm2m_client.c). I don't like this, but i had avoid to allocate a buffer inside peerinfo. Is there a problem to allocate it there?

Comment (rlubos) [reply include/net/lwm2m.h:101]: The point is you don't need to allocate - just copy the pointer. The URL is either stored in the Security or Firmware Update object implementation so we just need to copy the location. Current approach won't work if you for instance run bootstrap, and the bootstrap server sets the new URL - you'll still use the BS server hostname if you set this in the application.

**Pull request closed without merging**